File: equisense/frontend/lib/adapters/swingPipeline.ts
import type {
  GateAData, GateAResult, GateBInput, GateBResult,
  GateStatus, GateVerdict, RRInput, RRResult,
  StockType, TimeStopResult, SwingFinalResult, FinalVerdict,
} from '@/types'

// ── 하드코딩 상수 (공표값 변경 시 여기만 수정) ──────────
export const MACRO_CONSTANTS = {
  rate_bp: 0 as number,           // BOK 기준금리 동결 (기준: 2025-01)
  pmi: 51.2 as number,            // Korea Mfg PMI (기준: 2026-05)
  pmi_direction: 'up' as const,
  last_updated: '2026-05',
}

// ── Gate A ───────────────────────────────────────────────
export function checkGateA(data: GateAData): GateAResult {
  const axes: Record<string, GateStatus> = {}

  axes.vix = data.vix == null ? 'GO'
    : data.vix <= 20 ? 'GO' : data.vix <= 30 ? 'WARN' : 'STOP'

  axes.rate = data.rate_bp <= 25 ? 'GO' : data.rate_bp <= 50 ? 'WARN' : 'STOP'

  axes.pmi = data.pmi >= 50 ? 'GO'
    : data.pmi >= 45
      ? (data.pmi_direction === 'down' ? 'WARN' : 'GO')
      : 'STOP'

  if (data.kospi_price != null && data.kospi_ma200 != null) {
    axes.index = data.kospi_price > data.kospi_ma200 ? 'GO' : 'WARN'
  } else {
    axes.index = 'GO'
  }

  if (data.usdkrw != null) {
    axes.usdkrw = data.usdkrw >= 1400 ? 'WARN' : 'GO'
  }

  const verdict: GateVerdict = Object.values(axes).some(v => v === 'STOP') ? 'BLOCK' : 'PASS'
  return { data, axes, verdict }
}

// ── Gate B ───────────────────────────────────────────────
export function checkGateB(input: GateBInput): GateBResult {
  const layer1: GateStatus =
    input.market_foreign_days >= 3 && input.market_institution === 'buy' ? 'GO' :
    input.market_foreign_days <= -3 && input.market_institution === 'sell' ? 'STOP' : 'WARN'

  const layer2: GateStatus =
    input.sector_etf_days >= 5 ? 'GO' :
    input.sector_etf_days <= -3 ? 'STOP' : 'WARN'

  const short_ok = input.short_ratio < 0.03 && input.short_trend === 'decrease'
  const stock_ok = input.stock_foreign_days >= 5 && input.stock_institution_weeks >= 3
  const layer3: GateStatus =
    (stock_ok && short_ok) ? 'GO' :
    (input.stock_foreign_days <= -3 || input.stock_institution_weeks <= -2) ? 'STOP' : 'WARN'

  const verdict: GateVerdict =
    [layer1, layer2, layer3].some(l => l === 'STOP') ? 'BLOCK' : 'PASS'

  const matrix =
    layer2 === 'GO' && layer3 === 'GO' ? 'STRONG_BUY' :
    layer2 === 'GO' && layer3 !== 'GO' ? 'FIND_ALTERNATIVE' :
    layer2 !== 'GO' && layer3 === 'GO' ? 'HEADWIND_SHORT_ONLY' : 'NO_ENTRY'

  return { input, layer1, layer2, layer3, matrix, verdict }
}

// ── R:R 검증 ─────────────────────────────────────────────
export function checkRR(input: RRInput): RRResult {
  const { entry, stop, target } = input
  if (stop >= entry || target <= entry) {
    return { rr: 0, loss_pct: 0, gain_pct: 0, breakeven_winrate: 100, verdict: 'BLOCK' }
  }
  const risk   = entry - stop
  const reward = target - entry
  const rr     = reward / risk
  return {
    rr:                 Math.round(rr * 100) / 100,
    loss_pct:           Math.round((entry - stop) / entry * 1000) / 10,
    gain_pct:           Math.round((target - entry) / entry * 1000) / 10,
    breakeven_winrate:  Math.round(1 / (1 + rr) * 1000) / 10,
    verdict:            rr >= 2.0 ? 'PASS' : rr >= 1.5 ? 'CAUTION' : 'BLOCK',
  }
}

// ── 시간 손절 ────────────────────────────────────────────
const TRADING_DAYS: Record<StockType, number> = {
  high_beta: 10,
  value:     15,
  small_cap: 10,
}

function addTradingDays(date: Date, n: number): Date {
  const d = new Date(date)
  let added = 0
  while (added < n) {
    d.setDate(d.getDate() + 1)
    if (d.getDay() !== 0 && d.getDay() !== 6) added++
  }
  return d
}

function tradingDaysBetween(a: Date, b: Date): number {
  if (a > b) return -tradingDaysBetween(b, a)
  let count = 0
  const cur = new Date(a)
  while (cur < b) {
    cur.setDate(cur.getDate() + 1)
    if (cur.getDay() !== 0 && cur.getDay() !== 6) count++
  }
  return count
}

export function getTimeStop(entryDate: Date, stockType: StockType): TimeStopResult {
  const totalDays = TRADING_DAYS[stockType]
  const deadline  = addTradingDays(entryDate, totalDays)
  const today     = new Date()
  const elapsed   = Math.max(0, tradingDaysBetween(entryDate, today))
  const remaining = Math.max(0, totalDays - elapsed)

  const status: TimeStopResult['status'] =
    remaining > 5 ? 'HOLDING' : remaining > 0 ? 'PREPARE_EXIT' : 'TIME_STOP'

  const action =
    status === 'HOLDING'       ? `보유 유지 — ${remaining}거래일 남음` :
    status === 'PREPARE_EXIT'  ? `청산 준비 — ${remaining}거래일 남음` :
    '시간 손절 실행 — 익일 시초가 청산'

  return {
    entry_date: entryDate.toISOString().slice(0, 10),
    deadline:   deadline.toISOString().slice(0, 10),
    total_days: totalDays,
    elapsed,
    remaining,
    status,
    action,
  }
}

// ── 최종 판정 ────────────────────────────────────────────
export function getFinalVerdict(
  gateA: GateVerdict,
  gateB: GateVerdict,
  step1Pass: boolean,
  rr: RRResult,
  entry: number,
  stop: number,
  target: number,
  stockType: StockType = 'high_beta',
): SwingFinalResult {
  const today = new Date()
  const timeStop = getTimeStop(today, stockType)

  let verdict: FinalVerdict
  let summaryLine: string

  if (gateA === 'BLOCK') {
    verdict = 'BLOCK'
    summaryLine = '거시환경 진입 불가 — 전면 대기'
  } else if (gateB === 'BLOCK') {
    verdict = 'BLOCK'
    summaryLine = '수급 진입 불가 — 외국인·기관 동반 이탈 확인'
  } else if (!step1Pass) {
    verdict = 'BLOCK'
    summaryLine = '재무 체력 미달 — 스윙 진입 부적합'
  } else if (rr.verdict === 'BLOCK') {
    verdict = 'BLOCK'
    summaryLine = `R:R ${rr.rr} : 1 — 기준 2:1 미달`
  } else if (gateB === 'PASS') {
    verdict = 'PASS'
    const fmt = (n: number) => n.toLocaleString('ko-KR')
    summaryLine = `진입 ${fmt(entry)}원 / 손절 ${fmt(stop)}원 / 목표 ${fmt(target)}원 / R:R ${rr.rr}:1`
  } else {
    verdict = 'CONDITIONAL'
    summaryLine = 'Gate B 수급 재확인 후 진입 결정'
  }

  return {
    verdict,
    gate_a: gateA,
    gate_b: gateB,
    step1_pass: step1Pass,
    rr,
    time_stop: timeStop,
    summary_line: summaryLine,
  }
}

File: equisense/my-skills/extracted/swing_pipeline.py
"""
swing_pipeline.py — 스윙 투자 7단계 완결형 파이프라인

사용법 (한미반도체 예시):
    python swing_pipeline.py \
        --ticker 042700 \
        --vix 18.5 --rate_bp 0 --pmi 52.3 --pmi_dir up \
        --idx_200ma above_up --usdkrw 1380 \
        --mkt_foreign_days 4 --mkt_institution buy \
        --sector_etf_days 6 \
        --stk_foreign_days 5 --stk_inst_weeks 3 \
        --short_ratio 0.02 --short_trend decrease \
        --total_debt 3200 --equity 7800 \
        --op_income 3100 --interest_exp 80 --ocf 2600 \
        --entry 68000 --stop 65960 --target 82000 \
        --stock_type high_beta
"""

import argparse
import json
from datetime import date

from gate_a import check_gate_a
from gate_b import check_gate_b
from steps  import (
    check_health_filter_manual,
    check_rr,
    get_time_stop,
)

SEPARATOR = "─" * 52


def fmt_status(s: str) -> str:
    icons = {'PASS': '✅', 'BLOCK': '🚫', 'GO': '✅',
             'WARN': '⚠️', 'STOP': '🚫'}
    return f"{icons.get(s, '')} {s}"


def run_pipeline(args) -> dict:
    report = {'ticker': args.ticker, 'run_date': date.today().isoformat()}

    print(f"\n{SEPARATOR}")
    print(f"  스윙 투자 진입 판정 — {args.ticker}")
    print(SEPARATOR)

    # ── Gate A ──────────────────────────────
    print("\n[Gate A] 거시환경")
    gate_a = check_gate_a(
        vix=args.vix,
        rate_change_bp=args.rate_bp,
        pmi=args.pmi,
        pmi_direction=args.pmi_dir,
        idx_vs_200ma=args.idx_200ma,
        usdkrw=args.usdkrw,
    )
    report['gate_a'] = gate_a
    for k in ('vix', 'rate', 'pmi', 'index'):
        if k in gate_a:
            print(f"  {k:10s}: {fmt_status(gate_a[k])}")
    print(f"  → {fmt_status(gate_a['gate_a'])}")

    if gate_a['gate_a'] == 'BLOCK':
        report['final'] = 'BLOCK'
        report['reason'] = f"Gate A 차단 — 대기 축: {gate_a['stop_axes']}"
        _print_final(report)
        return report

    # ── Gate B ──────────────────────────────
    print("\n[Gate B] 수급")
    gate_b = check_gate_b(
        market_foreign_days=args.mkt_foreign_days,
        market_institution=args.mkt_institution,
        sector_etf_days=args.sector_etf_days,
        stock_foreign_days=args.stk_foreign_days,
        stock_institution_weeks=args.stk_inst_weeks,
        short_ratio=args.short_ratio,
        short_trend=args.short_trend,
    )
    report['gate_b'] = gate_b
    for k in ('layer1', 'layer2', 'layer3'):
        print(f"  {k:10s}: {fmt_status(gate_b[k])}")
    print(f"  매트릭스 : {gate_b['matrix']} — {gate_b['matrix_action']}")
    print(f"  → {fmt_status(gate_b['gate_b'])}")

    if gate_b['gate_b'] == 'BLOCK':
        report['final'] = 'BLOCK'
        report['reason'] = f"Gate B 차단 — 차단 레이어: {gate_b['stop_layers']}"
        _print_final(report)
        return report

    # ── Step 1: 체력 필터 ───────────────────
    print("\n[Step 1] 체력 필터")
    step1 = check_health_filter_manual(
        total_debt=args.total_debt,
        equity=args.equity,
        operating_income=args.op_income,
        interest_expense=args.interest_exp,
        operating_cash_flow=args.ocf,
    )
    report['step1'] = step1
    print(f"  부채비율   : {step1['debt_ratio']}% → {fmt_status('GO' if step1['debt_pass'] else 'STOP')}")
    print(f"  이자보상배율: {step1['icr']}× → {fmt_status('GO' if step1['icr_pass'] else 'STOP')}")
    print(f"  OCF       : {step1['ocf']:,.0f}억 → {fmt_status('GO' if step1['ocf_pass'] else 'STOP')}")

    if not step1['step1_pass']:
        report['final'] = 'BLOCK'
        report['reason'] = 'Step 1 체력 필터 미통과'
        _print_final(report)
        return report

    # ── Step 2~4: 수동 확인 안내 ────────────
    print("\n[Step 2] 촉발 이벤트  → HTS·DART 수동 확인")
    print("  어닝 서프라이즈 +5%↑ / QoQ 이익률 2분기 개선 / 가이던스 상향")
    print("\n[Step 3] 섹터·종목 수급  → Gate B 결과 재확인")
    print(f"  {gate_b['matrix']} — {gate_b['matrix_action']}")
    print("\n[Step 4] 기술적 진입  → 차트 직접 확인")
    print("  눌림목: 20MA 근처 + 거래량 수렴 + RSI 40~60")
    print("  돌  파: 거래량 2배↑ + 종가 기준 저항선 돌파")

    # ── Step 5: R:R 검증 ────────────────────
    print("\n[Step 5] R:R 검증")
    step5 = check_rr(entry=args.entry, stop=args.stop, target=args.target)
    report['step5'] = step5
    print(f"  진입가  : {args.entry:,.0f}원")
    print(f"  손절선  : {args.stop:,.0f}원  ({step5['loss_pct']}%)")
    print(f"  목표가  : {args.target:,.0f}원  (+{step5['gain_pct']}%)")
    print(f"  R:R     : {step5['rr']} : 1")
    print(f"  손익분기 승률: {step5['breakeven_winrate']}%")
    print(f"  → {fmt_status(step5['verdict'])}")

    if step5['verdict'] == 'BLOCK':
        report['final'] = 'BLOCK'
        report['reason'] = f"R:R {step5['rr']} : 1 — 기준 미달 (최소 2.0 필요)"
        _print_final(report)
        return report

    # ── Step 6: 시간 손절 기한 ──────────────
    print("\n[Step 6] 시간 손절 기한")
    step6 = get_time_stop(date.today(), args.stock_type)
    report['step6'] = step6
    print(f"  종목 유형 : {args.stock_type}")
    print(f"  보유 기한 : {step6['deadline']} ({step6['total_days']}거래일)")
    print(f"  상태      : {step6['action']}")

    # ── 최종 판정 ───────────────────────────
    report['final'] = 'PASS'
    report['summary'] = {
        'entry':    args.entry,
        'stop':     args.stop,
        'target':   args.target,
        'rr':       step5['rr'],
        'deadline': step6['deadline'],
    }
    _print_final(report)
    return report


def _print_final(report: dict):
    print(f"\n{SEPARATOR}")
    if report['final'] == 'PASS':
        s = report['summary']
        print(f"  최종 판정: ✅ PASS — 전 단계 통과")
        print(f"  진입가 {s['entry']:,.0f}원 / 손절 {s['stop']:,.0f}원 / "
              f"목표 {s['target']:,.0f}원")
        print(f"  R:R {s['rr']} : 1  |  시간 손절 {s['deadline']}")
    else:
        print(f"  최종 판정: 🚫 BLOCK — {report.get('reason', '')}")
    print(SEPARATOR)


def main():
    p = argparse.ArgumentParser(description='스윙 투자 7단계 진입 판정')

    # 종목
    p.add_argument('--ticker',         required=True, help='종목코드 (예: 042700)')
    p.add_argument('--stock_type',     default='value',
                   choices=['high_beta', 'value', 'small_cap'])

    # Gate A
    p.add_argument('--vix',            type=float, default=18.0)
    p.add_argument('--rate_bp',        type=float, default=0)
    p.add_argument('--pmi',            type=float, default=51.0)
    p.add_argument('--pmi_dir',        default='up', choices=['up', 'down'])
    p.add_argument('--idx_200ma',      default='above_up',
                   choices=['above_up','above_down','below_up','below_down'])
    p.add_argument('--usdkrw',         type=float, default=0)

    # Gate B
    p.add_argument('--mkt_foreign_days',   type=int,   default=3)
    p.add_argument('--mkt_institution',    default='buy',
                   choices=['buy','sell','neutral'])
    p.add_argument('--sector_etf_days',    type=int,   default=5)
    p.add_argument('--stk_foreign_days',   type=int,   default=5)
    p.add_argument('--stk_inst_weeks',     type=int,   default=3)
    p.add_argument('--short_ratio',        type=float, default=0.02)
    p.add_argument('--short_trend',        default='decrease',
                   choices=['decrease','increase','stable'])

    # Step 1
    p.add_argument('--total_debt',     type=float, required=True, help='총부채(억원)')
    p.add_argument('--equity',         type=float, required=True, help='자기자본(억원)')
    p.add_argument('--op_income',      type=float, required=True, help='영업이익(억원)')
    p.add_argument('--interest_exp',   type=float, required=True, help='이자비용(억원)')
    p.add_argument('--ocf',            type=float, required=True, help='영업현금흐름(억원)')

    # Step 5
    p.add_argument('--entry',          type=float, required=True, help='진입가(원)')
    p.add_argument('--stop',           type=float, required=True, help='손절선(원)')
    p.add_argument('--target',         type=float, required=True, help='목표가(원)')

    args = p.parse_args()
    result = run_pipeline(args)

    with open(f"swing_{args.ticker}_{date.today().isoformat()}.json", 'w',
              encoding='utf-8') as f:
        json.dump(result, f, ensure_ascii=False, indent=2)
    print(f"\n결과 저장: swing_{args.ticker}_{date.today().isoformat()}.json")


if __name__ == '__main__':
    main()
